### ISSUE TITLE
Multimodel AlexNet is (may be) broken

### ISSUE THREAD

**Issue by samadejacobs** [LLNL/lbann#139]
I am getting about 1% and 8% top1 and top5 accuracy respectively after 10 epoch. Single AlexNet works fine ( ~81% top 5 accuracy @10 epoch).  For the multimodel, I had 16 models with 4 procs per model on Ray. Nothing else changed in prototext. I suspect that the minibatch size may have to change when we change from single to multimodel? Do we need to do that i.e. divide minibatch size by number of models? What other hyperparameters need to change?

**Comment by ndryden** [comment]
I believe (@bvanessen can correct me if I'm wrong) that the mini-batch size we specify is a per-model mini-batch size. If you want to keep the training dynamics roughly the same, you'll need to keep the global mini-batch size constant. AlexNet will not train well with a larger mini-batch size (unless you use e.g. my variable-size mini-batch stuff).

**Comment by samadejacobs** [comment]
Maintaining a fixed global mini-batch ( local mini-batch of 16 , global 256 for 16 models) results in max top5 accuracy of ~21% which is still less than ~81% achieved by single model.

**Comment by samadejacobs** [comment]
Also, if multi-model is doing mini-batch summation, it may (in principle) not work for LTFB as expected. Taken the AlexNet for example, number of trainers (models) will be limited to 256 at extreme case ( per model(trainer) mini-batch size of 1).

**Comment by ndryden** [comment]
You can disable the summation by disabling the imcomm callback. It doesn't make sense to use it when using ltfb.

**Comment by samadejacobs** [comment]
Yes, I did that already but still with poor learning. I suspect the real (or first) problem is with accuracy gaps between single and multi models, fixing that may fix part of the problem with LTFB.

**Comment by bvanessen** [comment]
The model should be changed so that the data set does not span multiple models.  Combining this with the removal of the immcom callback should fix the problem.

Brian C. Van Essen
vanessen1@llnl.gov
(w) 925-422-9300
(c) 925-290-5470

**Comment by timmoon10** [comment]
I get the following results on AlexNet when I train with multiple models (effective mini-batch size is always 256; standard inter-model communicator.)

| # Models | Epoch 1 Average Objective | Epoch 5 Average Objective | Epoch 10 Average Objective |
|----|----|----|----|
| 1 | 8.39 | 6.83 | 6.19 |
| 2 | 7.64 | 6.72 | 6.41 |
| 4 | 7.71 | 6.86 | 6.41 |
| 8 | 7.88 | 7.11 | 6.75 |

My gradient checks come back clean in the single-model case, so I suspect the discrepancy is from floating point error in the allreduces. I'll rerun these experiments with doubles.

**Comment by timmoon10** [comment]
Here are the results of the experiment with doubles (effective mini-batch size is always 256; standard inter-model communicator; GPU disabled; debug build)

| # Models | Epoch 1 Average Objective |
|----|----|
| 1 | 7.63 |
| 2 | 7.77 |
| 4 | 7.91 |
| 8 | 7.86 |

**Comment by bvanessen** [comment]
Have you setup the initialization so that it is using the sequentially consistent version?  When that is used the results should not change with the number of models.


Brian C. Van Essen
vanessen1@llnl.gov
(w) 925-422-9300
(c) 925-290-5470
---
Sent from VMware Boxer

On December 4, 2017 at 5:21:15 PM PST, Tim Moon <notifications@github.com> wrote:

Here are the results of the experiment with doubles (effective mini-batch size is always 256; standard inter-model communicator; GPU disabled; debug build)

# Models        Epoch 1 Average Objective
1       7.63
2       7.77
4       7.91
8       7.86

**Comment by ndryden** [comment]
If vanilla inter-model communication is used (i.e. not quantization), the models (i.e. the weights) should be identical up to floating point associativity/roundoff issues, otherwise we are not doing data-parallel training. This worked at one point.

Sequential consistency should be an issue only if we vary the number of processes used in each model.

**Comment by bvanessen** [comment]
The standard inter-model communication and summation is fixed in 20e1daf77e792c0799ece39123d371190009d083.

**Comment by timmoon10** [comment]
Running the test_io_buffers unit test, models produce identical results if the global mini-batch size is kept consistent. There are some small discrepancies in the multi-model case, but this is likely due to inconsistent computation of weighted averages of the objective function and metrics.
